### PR TITLE
ARROW-18056: [Ruby] Add support for building Arrow::Table from {name: Arrow::Tensor}

### PR DESCRIPTION
**Pull request by kou**
New #to_arrow_array protocol is also introduced.

### CODE REVIEW

**Comment by kou** [comment]
@mrkn What do you think about this approach?

File: ruby/red-arrow/lib/arrow/raw-table-converter.rb [hunk 1]
```diff
@@ -35,7 +35,11 @@ def convert
         fields = []
         @values = []
         @raw_table.each do |name, array|
-          array = ArrayBuilder.build(array) if array.is_a?(::Array)
+          if array.respond_to?(:to_arrow_array)
+            array = array.to_arrow_array
+          elsif array.is_a?(::Array)
```
Comment (mrkn) [inline ruby/red-arrow/lib/arrow/raw-table-converter.rb:40]: Do you think how to handle objects that respond to `to_ary` in addition to normal arrays?

**Review by mrkn — COMMENTED** [review]
@kou I like this approach.  I left a comment about objects that respond to `to_ary`.

File: ruby/red-arrow/lib/arrow/raw-table-converter.rb (hunk 1 above)
Comment (kou) [reply ruby/red-arrow/lib/arrow/raw-table-converter.rb]: It's a good idea.
I've added support for `#to_ary`.

Comment (mrkn) [reply ruby/red-arrow/lib/arrow/raw-table-converter.rb]: Ok. LGTM!

**Pull request merged**